### PR TITLE
Redesign of Firmware section at Setup tab

### PR DESCRIPTION
**Pull request by HThuren**
Changed according to request at discord
New design:
image

With local build and still logic if not online:
image

Press on Options give popup (possible to select text):
image

Old design:
image

### CODE REVIEW

**Comment by ctzsnooze** [comment]
Looks really great, @HThuren !

The orange buttons are very intuitive and helpful.

Could the `Defines Used` button, currently a grey 'hover' button, be also an orange 'click to open in new window' button?  I think it would be more intuitive like that.  And easier to copy and paste the defines used, if needed.

**Comment by HThuren** [comment]
@ctzsnooze thank you.
To change Options defined to orange button, a good idea, will require a popup window. Configurator only has such in preset tabs, and here quite sophiscated. 
I'll construct a more 'general' popup with title, body and OK button. This can be in use in other future functionality.

**Comment by nerdCopter** [comment]
IMHO, i prefer hover-over, not pop-out. sorry if i'm the only one.
I also prefer old tooltip-hover for description, not the new always-active text.

**Comment by haslinghuis** [comment]
Hold fire on this PR as we are working on other methods to retrieve build options from firmware without the need for Build API queries as we need the options for LUA too.

**Comment by HThuren** [comment]
right, anyway changed back to help in cf_help tooltip, as wanted from @nerdCopter, since the yellow note require lot of space.
Hover over are a possibility, but did look into a popup window a a general tool.
But halt both until @haslinghuis give green go

**Comment by haslinghuis** [comment]
Just saw it yesterday as it's not my contribution - so improving this PR now does not hurt as we only attend to improve back-end call functions.

**Comment by ctzsnooze** [comment]
If we keep the 'hover' behaviour, we should use the standard `?` icon that we use for other hover elements.

A 'greyed out' button that doesn't work implies a 'normal' button that is disabled.  It is bad UI design.  The normal user might try to click it, just in case the colour loss was a mistake, but nothing happens, so that confirms their understanding that the button is disabled.  They are likely to report this as a bug.

We don't have any other 'greyed out' button elements that mean 'hover over me' on any other windows.

A pop-out lets us copy the text, whereas a hover doesn't.

Anyhow, I still prefer an orange button with a pop-out, but if you want to keep it a hover, then use `?` as the UI element, since people intuitively will hover over those.

PS There is another issue with the grey button hover, it is very slow to open, much slower than hovering over a `?` element, which is nearly instantaneous.  Hovers need to be instant otherwise they appear broken.

**Comment by HThuren** [comment]
> If we keep the 'hover' behaviour, we should use the standard `?` icon that we use for other hover elements.
> 
> A 'greyed out' button that doesn't work implies a 'normal' button that is disabled. It is bad UI design. The normal user might try to click it, just in case the colour loss was a mistake, but nothing happens, so that confirms their understanding that the button is disabled. They are likely to report this as a bug.
> 
> We don't have any other 'greyed out' button elements that mean 'hover over me' on any other windows.
> 
> A pop-out lets us copy the text, whereas a hover doesn't.
> 
> Anyhow, I still prefer an orange button with a pop-out, but if you want to keep it a hover, then use `?` as the UI element, since people intuitively will hover over those.
> 
> PS There is another issue with the grey button hover, it is very slow to open, much slower than hovering over a `?` element, which is nearly instantaneous. Hovers need to be instant otherwise they appear broken.

I agree, the grey hover are different behaviour than elsewhere in BF.
I go for 3 orange buttons, 
image

But are not done yet with at popup for "Options defined"

**Comment by nerdCopter** [comment]
@HThuren , just a note, i'm a relative new-comer to betaflight team. my opinions should be regarded as my own, not the team's.  other team members should have priority and say as well.  that said, i've been on computers since 1985 and i think i have good feel for things.

**Comment by ctzsnooze** [comment]
where are we at?

**Comment by HThuren** [comment]
Maybe part of 10.10 ?

File: src/js/popup.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import 'jqueryPlugins';
+import GUI from "gui";
```
Comment (haslinghuis) [inline src/js/popup.js:2]: This import is not being used.
```suggestion
```

File: src/js/tabs/setup.js [hunk 2]
```diff
@@ -477,6 +499,9 @@ setup.initialize = function (callback) {
             if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_46) && FC.CONFIG.cpuTemp) {
                 cputemp_e.html(`${FC.CONFIG.cpuTemp.toFixed(0)} &#8451;`);
             }
+            else {
```
Comment (haslinghuis) [inline src/js/tabs/setup.js:502]:
```suggestion
            } else {
```

Comment (HThuren) [reply src/js/tabs/setup.js:502]: Good catch

File: src/js/popup.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        this._domDialog = domDialog;
+        this._sourceSelectedPromiseResolve = null;
+    }
+
+    load() {
+        return new Promise(resolve => {
+            this._domDialog.load("./tabs/popup.html",
+            () => {
+                this._setupDialog();
+                this._initializeSources();
```
Comment (haslinghuis) [inline src/js/popup.js:16]: Seems leftover :)
```suggestion
```

**Comment by haslinghuis** [comment]
This looks very good. Only remark left is the presentation:

image

**Comment by HThuren** [comment]
Yes, the presentation are raw, what do we like here , some columns 3 x 3 or ?

File: src/js/popup.js (hunk 3 above)
Comment (HThuren) [reply src/js/popup.js]: popup.js itself was a leftrover, sorry, no removed

File: src/js/tabs/setup.js [hunk 4]
```diff
@@ -416,35 +415,58 @@ setup.initialize = function (callback) {
...
 
-            if (FC.CONFIG.buildOptions.length) {
-                let buildOptions = "";
-                build_opt_e.text = "";
+                if (FC.CONFIG.buildOptions.length) {
 
-                for (const buildOption of FC.CONFIG.buildOptions) {
-                    buildOptions = `${buildOptions} &nbsp ${buildOption}`;
+                    for (const buildOptionElement of FC.CONFIG.buildOptions) {
+                        buildOptionList = `${buildOptionList} &nbsp ${buildOptionElement}`;
```
Comment (haslinghuis) [inline src/js/tabs/setup.js:443]:
```suggestion
                        buildOptionList = `${buildOptionList} ${buildOptionElement}`;
```

would present it like
image

Doing 3 x 3 would be even better. Could use  `.padEnd()`

File: src/js/tabs/setup.js [hunk 5]
```diff
@@ -416,35 +415,58 @@ setup.initialize = function (callback) {
...
+                    const buildConfig = `<span class="buildInfoBtn" title="${i18n.getMessage('initialSetupInfoBuildConfig')}: ${buildRoot}/json">
+                                         <a href="${buildRoot}/json" target="_blank"><strong>${i18n.getMessage('initialSetupInfoBuildConfig')}</strong></a></span>`;
+
+                    const buildLog =    `<span class="buildInfoBtn" title="${i18n.getMessage('initialSetupInfoBuildLog')}: ${buildRoot}/log">
+                                         <a href="${buildRoot}/log" target="_blank"><strong>${i18n.getMessage('initialSetupInfoBuildLog')}</strong></a></span>`;
+
+                    const buildOptions = `<span class="buildInfoBtn" title="${i18n.getMessage('initialSetupInfoBuildOptionList')}">
+                                         <a class="buildOptions disabled" href=#"><strong>${i18n.getMessage('initialSetupInfoBuildOptions')}</strong></a></span>`;
+
+                    build_info_e.html(`${buildConfig} ${buildLog} &nbsp &nbsp ${buildOptions}`);
```
Comment (haslinghuis) [inline src/js/tabs/setup.js:458]: I believe &nbsp; should end with semicolon:
```suggestion
                    build_info_e.html(`${buildConfig} ${buildLog} &nbsp; &nbsp; ${buildOptions}`);
```

File: src/js/tabs/setup.js (hunk 4 above)
Comment (haslinghuis) [reply src/js/tabs/setup.js:443]: Hmm no .padEnd() does not seem to work with html as it removes the spaces :(

Comment (HThuren) [reply src/js/tabs/setup.js:443]: I try some loops

Comment (haslinghuis) [reply src/js/tabs/setup.js:443]: Might need css - flex columns for that as fonts are disproportional.

File: src/main.html [hunk 6]
```diff
@@ -37,6 +37,7 @@
     <link type="text/css" rel="stylesheet" href="./css/tabs/osd.css" media="all"/>
     <link type="text/css" rel="stylesheet" href="./css/tabs/vtx.css" media="all"/>
     <link type="text/css" rel="stylesheet" href="./css/tabs/power.css" media="all"/>
+    <link type="text/css" rel="stylesheet" href="./css/tabs/popupDialog.css" media="all"/>
```
Comment (haslinghuis) [inline src/main.html:40]: This would no longer be needed also?

File: src/tabs/setup.html [hunk 7]
```diff
@@ -251,6 +247,9 @@
         </div>
     </div>
 
+    <dialog id="popup_sources_dialog">
+    </dialog>
```
Comment (haslinghuis) [inline src/tabs/setup.html:251]: This also seems not being used.

Comment (HThuren) [reply src/tabs/setup.html:251]: yes, now removed

File: src/main.html (hunk 6 above)
Comment (HThuren) [reply src/main.html:40]: yrk, I spend hours to make a class for general purpose, but that was too much.
Sorry for the leftover

**Comment by haslinghuis** [comment]
@HThuren look at this:

image

**Comment by haslinghuis** [comment]
In setup.less around line 188 add
```
diff --git a/src/css/tabs/setup.less b/src/css/tabs/setup.less
index 78a65008..1550b827 100644
--- a/src/css/tabs/setup.less
+++ b/src/css/tabs/setup.less
@@ -188,6 +188,17 @@
                margin-left: 99px;
        }
 }
.grid-container {
       display: grid;
       grid-template-columns: auto auto auto;
       padding: 10px;
}
.grid-item {
       background-color: rgba(255, 255, 255, 0.8);
       padding: 10px;
       text-align: center;
}
```
In setup.js around line 438 change to
```
if (FC.CONFIG.buildOptions.length) {
    buildOptionList = `<div class="grid-container">`;
    for (const buildOptionElement of FC.CONFIG.buildOptions) {
        buildOptionList += `<div class="grid-item">${buildOptionElement}</div>`;
    }
    buildOptionList += `</div>`;
}
```

**Comment by HThuren** [comment]
Yes, I was on to that, working to have it to center in frame.

**Comment by haslinghuis** [comment]
Solved in the snippet above ? `text-align: center;`

**Comment by HThuren** [comment]
yes and no :-)
image

**Comment by HThuren** [comment]
and added `user-select: text;`

File: src/js/tabs/setup.js [hunk 8]
```diff
@@ -416,35 +415,76 @@ setup.initialize = function (callback) {
...
+                    } else {
+                        grid += "<td></td>"; // empty cell, to ajust
+                    }
+                }
+                grid += "</tr>";
+            }
+            grid += "</table></div>";
+
+            return grid;
+        }
```
Comment (haslinghuis) [inline src/js/tabs/setup.js:451]: I prefer implementing my earlier suggestion instead. As the code is much leaner and as this is using tables while we are using grid layout does not make sense. The code provided is dynamic on itself and uses css for presentation.

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by HThuren** [comment]
ready !

image

File: src/js/tabs/setup.js (hunk 8 above)
Comment (haslinghuis) [reply src/js/tabs/setup.js:451]: See https://github.com/betaflight/betaflight-configurator/pull/3770#issuecomment-1938932576

Comment (HThuren) [reply src/js/tabs/setup.js:451]: Now get your idea, but without table I have only an one collum layout, missing empty cell `<td></td>` 
`<div class='${classItem}>&nbsp;</div>` or `<div>&nbsp;</div>` won't give the same
Do you know how to have an empty placeholder with `div` element

Comment (HThuren) [reply src/js/tabs/setup.js]: Now you get what you want :-)
image

**Review by ctzsnooze — APPROVED** [review]
Looks very nice

**Comment by HThuren** [comment]
@nerdCopter your images are not shown

**Comment by nerdCopter** [comment]
> @nerdCopter your images are not shown

:( it is your screenshot from your comment and haslinghuis's screen from his

**Comment by HThuren** [comment]
Right, the image in description of PR are updated.

**Review by nerdCopter — APPROVED** [review]
approving. tested both local and cloud hex; looks good :rocket:

**Comment by haslinghuis** [comment]
Personal don't like the borders and the need to use rows with grid layout. It should just use the grid layout with elements and css.

Found some other issues:

image

image

**Comment by haslinghuis** [comment]
image

**Comment by HThuren** [comment]
css need the row col construction, otherwise there was only one (long) col.

without border - can be OK - :
image

**Comment by haslinghuis** [comment]
Disagree:

With some CSS adjustments and this small code we get dynamic result

```
    if (FC.CONFIG.buildOptions.length) {
        buildOptions = `<div class="grid-container">`;
        for (const buildOptionElement of FC.CONFIG.buildOptions) {
            buildOptions += `<div class="grid-item">${buildOptionElement}</div>`;
        }
        buildOptions += `</div>`;
    }
```

image

**Comment by HThuren** [comment]
right, I give it at try :-)

**Comment by HThuren** [comment]
I don't like it:
image

better with and without border
image

**Comment by haslinghuis** [comment]
Can define two columns in the grid container:

image

**Comment by HThuren** [comment]
NOW done, use the more simple and dynamic grid. Support dark theme.

**Comment by nerdCopter** [comment]
`fd215243`
image

flow gets wonky at come widths, but not very important:
wide:
image
1202px: (odd)
image
1164px: (okay by me)
image

**Review by nerdCopter — APPROVED** [review]

File: src/js/tabs/setup.js (hunk 2 above)
Comment (haslinghuis) [reply src/js/tabs/setup.js:508]: This was not resolved :)

**Review by haslinghuis — APPROVED** [review]

**Pull request merged**